File: composeApp/src/androidMain/java/org/application/shikiapp/shared/CanvasSubtitleOutput.java
package org.application.shikiapp.shared;

import static org.application.shikiapp.shared.SubtitleView.DEFAULT_BOTTOM_PADDING_FRACTION;
import static org.application.shikiapp.shared.SubtitleView.DEFAULT_TEXT_SIZE_FRACTION;

import android.content.Context;
import android.graphics.Canvas;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.media3.common.text.Cue;
import androidx.media3.common.util.UnstableApi;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

@UnstableApi
final class CanvasSubtitleOutput extends View implements SubtitleView.Output {

  private final List<SubtitlePainter> painters;

  private List<Cue> cues;
  private @Cue.TextSizeType int textSizeType;
  private float textSize;
  private CaptionStyleCompat style;
  private float bottomPaddingFraction;

  public CanvasSubtitleOutput(Context context) {
    this(context, null);
  }

  public CanvasSubtitleOutput(Context context, @Nullable AttributeSet attrs) {
    super(context, attrs);
    painters = new ArrayList<>();
    cues = Collections.emptyList();
    textSizeType = Cue.TEXT_SIZE_TYPE_FRACTIONAL;
    textSize = DEFAULT_TEXT_SIZE_FRACTION;
    style = CaptionStyleCompat.DEFAULT;
    bottomPaddingFraction = DEFAULT_BOTTOM_PADDING_FRACTION;
  }

  @Override
  public void update(
          List<Cue> cues,
          CaptionStyleCompat style,
          float textSize,
          @Cue.TextSizeType int textSizeType,
          float bottomPaddingFraction) {
    this.cues = cues;
    this.style = style;
    this.textSize = textSize;
    this.textSizeType = textSizeType;
    this.bottomPaddingFraction = bottomPaddingFraction;
    while (painters.size() < cues.size()) {
      painters.add(new SubtitlePainter(getContext()));
    }
    invalidate();
  }

  @Override
  public void dispatchDraw(@NonNull Canvas canvas) {
    @Nullable List<Cue> cues = this.cues;
    if (Objects.requireNonNull(cues).isEmpty()) {
      return;
    }

    int rawViewHeight = getHeight();

    int left = getPaddingLeft();
    int top = getPaddingTop();
    int right = getWidth() - getPaddingRight();
    int bottom = rawViewHeight - getPaddingBottom();
    if (bottom <= top || right <= left) {
      return;
    }
    int viewHeightMinusPadding = bottom - top;

    float defaultViewTextSizePx =
            SubtitleViewUtils.resolveTextSize(
                    textSizeType, textSize, rawViewHeight, viewHeightMinusPadding);
    if (defaultViewTextSizePx <= 0) {
      return;
    }

    int cueCount = cues.size();
    for (int i = 0; i < cueCount; i++) {
      Cue cue = cues.get(i);
      if (cue.verticalType != Cue.TYPE_UNSET) {
        cue = repositionVerticalCue(cue);
      }
      float cueTextSizePx =
              SubtitleViewUtils.resolveTextSize(
                      cue.textSizeType, cue.textSize, rawViewHeight, viewHeightMinusPadding);
      SubtitlePainter painter = painters.get(i);
      painter.draw(
              cue,
              style,
              defaultViewTextSizePx,
              cueTextSizePx,
              bottomPaddingFraction,
              canvas,
              left,
              top,
              right,
              bottom);
    }
  }

  private static Cue repositionVerticalCue(Cue cue) {
    Cue.Builder cueBuilder =
            cue.buildUpon()
                    .setPosition(Cue.DIMEN_UNSET)
                    .setPositionAnchor(Cue.TYPE_UNSET)
                    .setTextAlignment(null);

    if (cue.lineType == Cue.LINE_TYPE_FRACTION) {
      cueBuilder.setLine(1.0f - cue.line, Cue.LINE_TYPE_FRACTION);
    } else {
      cueBuilder.setLine(-cue.line - 1f, Cue.LINE_TYPE_NUMBER);
    }
    switch (cue.lineAnchor) {
      case Cue.ANCHOR_TYPE_END:
        cueBuilder.setLineAnchor(Cue.ANCHOR_TYPE_START);
        break;
      case Cue.ANCHOR_TYPE_START:
        cueBuilder.setLineAnchor(Cue.ANCHOR_TYPE_END);
        break;
      case Cue.ANCHOR_TYPE_MIDDLE:
      case Cue.TYPE_UNSET:
      default:
    }
    return cueBuilder.build();
  }
}

File: composeApp/src/androidMain/kotlin/org/application/shikiapp/shared/utils/ui/subtitles/CanvasSubtitleOutput.kt
package org.application.shikiapp.shared.utils.ui.subtitles

import android.content.Context
import android.graphics.Canvas
import android.util.AttributeSet
import android.view.View
import androidx.annotation.OptIn
import androidx.media3.common.text.Cue
import androidx.media3.common.util.UnstableApi
import org.application.shikiapp.shared.utils.ui.subtitles.SubtitleView.Companion.DEFAULT_BOTTOM_PADDING_FRACTION
import org.application.shikiapp.shared.utils.ui.subtitles.SubtitleView.Companion.DEFAULT_TEXT_SIZE_FRACTION

@UnstableApi
class CanvasSubtitleOutput @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : View(context, attrs), SubtitleView.Output {

    private val painters = mutableListOf<SubtitlePainter>()
    private var cues: List<Cue> = emptyList()
    @Cue.TextSizeType
    private var textSizeType: Int = Cue.TEXT_SIZE_TYPE_FRACTIONAL
    private var textSize: Float = DEFAULT_TEXT_SIZE_FRACTION
    private var style: CaptionStyleCompat = CaptionStyleCompat.DEFAULT
    private var bottomPaddingFraction: Float = DEFAULT_BOTTOM_PADDING_FRACTION

    override fun update(
        cues: List<Cue>,
        style: CaptionStyleCompat,
        defaultTextSize: Float,
        @Cue.TextSizeType defaultTextSizeType: Int,
        bottomPaddingFraction: Float
    ) {
        this.cues = cues
        this.style = style
        this.textSize = defaultTextSize
        this.textSizeType = defaultTextSizeType
        this.bottomPaddingFraction = bottomPaddingFraction

        while (painters.size < cues.size) {
            painters.add(SubtitlePainter(context))
        }
        invalidate()
    }

    @OptIn(UnstableApi::class)
    override fun dispatchDraw(canvas: Canvas) {
        val currentCues = cues
        if (currentCues.isEmpty()) return

        val rawViewHeight = height
        val currentLeft = paddingLeft
        val currentTop = paddingTop
        val currentRight = width - paddingRight
        val currentBottom = rawViewHeight - paddingBottom

        if (currentBottom <= currentTop || currentRight <= currentLeft) return

        val viewHeightMinusPadding = currentBottom - currentTop
        val defaultViewTextSizePx = SubtitleViewUtils.resolveTextSize(
            textSizeType, textSize, rawViewHeight, viewHeightMinusPadding
        )

        if (defaultViewTextSizePx <= 0) return

        currentCues.forEachIndexed { index, originalCue ->
            val cue = if (originalCue.verticalType != Cue.TYPE_UNSET) {
                repositionVerticalCue(originalCue)
            } else {
                originalCue
            }

            val cueTextSizePx = SubtitleViewUtils.resolveTextSize(
                cue.textSizeType, cue.textSize, rawViewHeight, viewHeightMinusPadding
            )

            val painter = painters[index]
            painter.draw(
                cue = cue,
                style = style,
                defaultTextSizePx = defaultViewTextSizePx,
                cueTextSizePx = cueTextSizePx,
                bottomPaddingFraction = bottomPaddingFraction,
                canvas = canvas,
                cueBoxLeft = currentLeft,
                cueBoxTop = currentTop,
                cueBoxRight = currentRight,
                cueBoxBottom = currentBottom
            )
        }
    }

    companion object {
        @OptIn(UnstableApi::class)
        private fun repositionVerticalCue(cue: Cue) = cue.buildUpon()
            .setPosition(Cue.DIMEN_UNSET)
            .setPositionAnchor(Cue.TYPE_UNSET)
            .setTextAlignment(null)
            .apply {
                if (cue.lineType == Cue.LINE_TYPE_FRACTION) {
                    setLine(1.0f - cue.line, Cue.LINE_TYPE_FRACTION)
                } else {
                    setLine(-cue.line - 1f, Cue.LINE_TYPE_NUMBER)
                }

                when (cue.lineAnchor) {
                    Cue.ANCHOR_TYPE_END -> lineAnchor = Cue.ANCHOR_TYPE_START
                    Cue.ANCHOR_TYPE_START -> lineAnchor = Cue.ANCHOR_TYPE_END
                }
            }
            .build()
    }
}